TFD System.IO.DirectoryNotFoundException error

Installing from the TFD CD on current release produces the below error if the .\Content\mod folder doesn't exist. Standard CD install handles this with a Directory.Create

    An unhandled exception of type 'System.IO.DirectoryNotFoundException' occurred in mscorlib.dll
    Additional information: Could not find a part of the path     
    'C:\Users\#UserName#\Documents\OpenRA\Content\ra\redalert.mix'.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27019922-tfd-system-io-directorynotfoundexception-error?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

I managed to fix this issue and get the installation running, (pushed to my fork)
although after installation i get:

Failed to initialize hardware cursors. Falling back to software cursors.
Error was: Cross-thread operation not valid: This method must be called from the same thread that created this object.
Exception of type `System.InvalidOperationException`: Cross-thread operation not valid: This method must be called from the same thread that created this object.
  at OpenRA.Platforms.Default.ThreadAffine.VerifyThreadAffinity () [0x00000] in <filename unknown>:0 
  at OpenRA.Platforms.Default.Sdl2GraphicsDevice.CreateTexture () [0x00000] in <filename unknown>:0 
  at OpenRA.Graphics.HardwarePalette..ctor () [0x00000] in <filename unknown>:0 
  at OpenRA.Graphics.SoftwareCursor..ctor (OpenRA.Graphics.CursorProvider cursorProvider) [0x00000] in <filename unknown>:0 
  at OpenRA.Game.InitializeMod (System.String mod, OpenRA.Arguments args) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.Logic.InstallLogic+<InstallLogic>c__AnonStorey0.<>m__0 () [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.Logic.InstallFromCDLogic+<InstallTFD>c__AnonStorey1.<>m__0 () [0x00000] in <filename unknown>:0 
  at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Cross-thread operation not valid: This method must be called from the same thread that created this object.
  at OpenRA.Platforms.Default.ThreadAffine.VerifyThreadAffinity () [0x00000] in <filename unknown>:0 
  at OpenRA.Platforms.Default.Sdl2GraphicsDevice.CreateTexture () [0x00000] in <filename unknown>:0 
  at OpenRA.Graphics.HardwarePalette..ctor () [0x00000] in <filename unknown>:0 
  at OpenRA.Graphics.SoftwareCursor..ctor (OpenRA.Graphics.CursorProvider cursorProvider) [0x00000] in <filename unknown>:0 
  at OpenRA.Game.InitializeMod (System.String mod, OpenRA.Arguments args) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.Logic.InstallLogic+<InstallLogic>c__AnonStorey0.<>m__0 () [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.Logic.InstallFromCDLogic+<InstallTFD>c__AnonStorey1.<>m__0

The assets install successfully and can be ran by reopening OpenRA, I am not sure whether this is coincidental or a fault in the Installation logic.

--- a/AUTHORS
+++ b/AUTHORS
@@ -65,6 +65,7 @@ Also thanks to:
     * Iran
     * Jacob Dufault (jacobdufault)
     * James Dunne (jsd)
+    * James Gilbert (DSUK)
     * Jan-Willem Buurlage (jwbuurlage)
     * Jason (atlimit8)
     * Jeff Harris (jeff_1amstudios)

--- a/OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs
+++ b/OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs
@@ -443,6 +443,7 @@ namespace OpenRA.FileSystem
 
 		public void ExtractFile(uint index, string fileName)
 		{
+			Directory.CreateDirectory(Path.GetDirectoryName(fileName));
 			using (var destfile = File.Open(fileName, FileMode.Create))
 				GetContentById(index, destfile);
 		}

--- a/OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromCDLogic.cs
+++ b/OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromCDLogic.cs
@@ -137,7 +137,11 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 					}
 				}
 
-				afterInstall();
+				Game.RunAfterTick(() =>
+				{
+					Ui.CloseWindow();
+					afterInstall();
+				});
 			}) { IsBackground = true }.Start();
 		}
 

